File: 0279-perfect-squares/0279-perfect-squares.py
from queue import Queue

class Solution:
    def numSquares(self, n: int) -> int:
        # Function to generate all perfect squares up to n
        def generateSquares(n):
            squares = []
            i = 1
            while i * i <= n:
                squares.append(i * i)
                i += 1
            return squares
        
        # Initialize a queue for BFS and a set to keep track of visited numbers
        queue = Queue()
        visited = set()
        
        # Start from n and add it to the queue with step count 0
        queue.put((n, 0))
        
        while not queue.empty():
            current, steps = queue.get()
            
            # Explore all possible perfect squares less than or equal to the current number
            for square in generateSquares(current):
                next_num = current - square
                
                # If the next number becomes 0, we have found a combination
                if next_num == 0:
                    return steps + 1
                
                # If the next number is not visited, add it to the queue with updated step count
                if next_num not in visited:
                    queue.put((next_num, steps + 1))
                    visited.add(next_num)
        
        # If no solution is found, this line should not be reached
        return -1

# Example usage:
solution = Solution()
result = solution.numSquares(12)
print(result)

File: 0279-perfect-squares/0279-perfect-squares.cpp
#include <vector>
#include <cmath>

class Solution {
public:
    int numSquares(int n) {
        // Create a vector to store the least number of perfect squares needed to sum to each number from 0 to n
        std::vector<int> dp(n + 1, INT_MAX);
        
        // Base case: 0 requires 0 perfect squares to sum up to
        dp[0] = 0;
        
        // Iterate from 1 to n
        for (int i = 1; i <= n; ++i) {
            // Iterate over all possible perfect squares that are less than or equal to i
            for (int j = 1; j * j <= i; ++j) {
                // Update the dp[i] if taking the current perfect square (j*j)
                // reduces the number of perfect squares needed to sum up to i
                dp[i] = std::min(dp[i], dp[i - j * j] + 1);
            }
        }
        
        // The least number of perfect squares needed to sum up to n is stored in dp[n]
        return dp[n];
    }
};
